application-admintools-default/src/main/java/com/xwiki/admintools/internal/data/AbstractDataProvider.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.admintools.internal.data;
+
+import java.io.StringWriter;
+import java.io.Writer;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.script.ScriptContext;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.phase.Initializable;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.script.ScriptContextManager;
+import org.xwiki.template.Template;
+import org.xwiki.template.TemplateManager;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xwiki.admintools.DataProvider;
+
+/**
+ * To be extended by {@link DataProvider} implementations to simplify the code.
+ *
+ * @version $Id$
+ * @since 1.0
+ */
+public abstract class AbstractDataProvider implements DataProvider, Initializable
+{
+    @Inject
+    protected TemplateManager templateManager;
+
+    @Inject
+    protected Provider<XWikiContext> xcontextProvider;
+
+    @Inject
+    protected Logger logger;
+
+    @Inject
+    private ScriptContextManager scriptContextManager;
+
+    @Override
+    public void initialize() throws InitializationException
+    {
+        // Overwrite to initialize a component
+    }
+
+    /**
+     * Generates the template of a data provider.
+     *
+     * @param data contains the data to be shown in the template.
+     * @param template path to the template.
+     * @param hint data provider identifier.
+     * @return String containing the generated template.
+     */
+    protected String getRenderedTemplate(String template, Map<String, String> data, String hint)
+    {
+        Writer writer = new StringWriter();
+        Template customTemplate = this.templateManager.getTemplate(template);
+        try {
+            // Binds the data provided to the template.
+            this.bindData(hint, data);
+            this.templateManager.render(customTemplate, writer);
```

Comment:
The render(templateName) method requires a String as a parameter: `String render(String templateName) throws Exception;`
I need to use a template as a parameter to be able to bind the data provided by the components. The solution that I found is to use `void render(Template template, Writer writer) throws Exception;` and a writer to add the resulting String,

Reply:
Yes, it needs the `templateName`. But the `writer` part is not mandatory.
There is a method signature [without the writer](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-template/xwiki-platform-template-api/src/main/java/org/xwiki/template/TemplateManager.java#L216) that will directly return the string, without you doing `writer.toString()`

Reply:
>  I need to use a Template object as a parameter because I need to bind the json provided by the data provider to the template.

No you don't.

What you do in bindData does not have anything to do with the Template object you get before calling it: what you put in the script context will be available during the template execution. `this.templateManager.render(customTemplate)` is doing the exact same thing as you: get the Template and then render it in a writer.